smithy-model/src/main/java/software/amazon/smithy/model/knowledge/TextInstance.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.knowledge;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.List;
+import java.util.Objects;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.utils.ListUtils;
+
+public final class TextInstance {
+    private final TextLocation locationType;
+    private final String text;
+    private final Shape shape;
+    private final Trait trait;
+    private final List<String> traitPropertyPath;
+
+    private TextInstance(Builder builder) {
```

Comment:
I wonder if we even need a builder here -- for example, could we use named factory methods? This would capture each invariant more explicitly.

```
TextInstance createNamespaceText(String namespace);
TextInstance createShapeInstance(Shape shape, anythingElse?);
TextInstance createTraitInstance(Shape shape, Trait trait, Collection<String> traitPath);
```

Reply:
Builders looked be the more common pattern in the rest of the code. After a second look, I see some factory creation methods as well, but most of those are using Builders right underneath. How about having both? Builder exists, along with the factory methods on top?